src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1013 @@
+    { MaterialAttributeType::Vector3, MaterialAttribute::EmissiveColor, MaterialAttribute::EmissiveTexture, MaterialAttribute::EmissiveTextureMatrix, MaterialAttribute::EmissiveTextureCoordinates, UFBX_MATERIAL_PBR_EMISSION_COLOR, UFBX_MATERIAL_PBR_EMISSION_FACTOR },
+    { (MaterialAttributeType)0, (MaterialAttribute)0, MaterialAttribute::NormalTexture, MaterialAttribute::NormalTextureMatrix, MaterialAttribute::NormalTextureCoordinates, UFBX_MATERIAL_PBR_NORMAL_MAP, -1 },
+    { (MaterialAttributeType)0, (MaterialAttribute)0, MaterialAttribute::OcclusionTexture, MaterialAttribute::OcclusionTextureMatrix, MaterialAttribute::OcclusionTextureCoordinates, UFBX_MATERIAL_PBR_AMBIENT_OCCLUSION, -1 },
+};
+
+}
+
+struct TextureBlob {
+    const void *data;
+    std::size_t size;
+};
+
+struct UfbxImporter::State {
+    ufbx_scene_ref scene;
+
+    Containers::Array<MeshChunk> meshChunks;
+
+    /* Mapping from ufbx_mesh.typed_id -> State.meshChunks */
+    Containers::Array<UnsignedInt> meshChunkBase;
+
+    UnsignedInt nodeIdOffset = 0;
+    UnsignedInt originalNodeCount = 0;
+    UnsignedInt nodeCountWithSynthetic = 0;
+
+    bool fromFile = false;
+
+    UnsignedInt imageImporterId = ~UnsignedInt{};
+    Containers::Optional<AnyImageImporter> imageImporter;
+};
+
+UfbxImporter::UfbxImporter() {
+}
+
+UfbxImporter::UfbxImporter(PluginManager::AbstractManager& manager, const Containers::StringView& plugin): AbstractImporter{manager, plugin} {
+}
+
+UfbxImporter::~UfbxImporter() = default;
+
+ImporterFeatures UfbxImporter::doFeatures() const { return ImporterFeature::OpenData|ImporterFeature::OpenState|ImporterFeature::FileCallback; }
+
+bool UfbxImporter::doIsOpened() const { return !!_state; }
+
+void UfbxImporter::doClose() { _state = nullptr; }
+
+void UfbxImporter::doOpenData(Containers::Array<char>&& data, const DataFlags) {
+    ufbx_load_opts opts = loadOptsFromConfiguration(configuration());
+    ufbx_error error;
+    ufbx_scene *scene = ufbx_load_memory(data.data(), data.size(), &opts, &error);
+    if (!scene) {
+        logError("Trade::UfbxImporter::openData(): loading failed: ", error);
+    }
+
+    /* Handle the opened scene with doOpenState() */
+    doOpenState(scene, {});
+    ufbx_free(scene);
+}
+
+void UfbxImporter::doOpenFile(Containers::StringView filename) {
+    ufbx_load_opts opts = loadOptsFromConfiguration(configuration());
+    ufbx_error error;
+    ufbx_scene *scene = ufbx_load_file_len(filename.data(), filename.size(), &opts, &error);
+    if (!scene) {
+        logError("Trade::UfbxImporter::openData(): loading failed: ", error);
+    }
+
+    /* Handle the opened scene with doOpenState() */
+    doOpenState(scene, filename);
+    ufbx_free(scene);
+}
+
+void UfbxImporter::doOpenState(const void* state, Containers::StringView path) {
```

Comment:
Yea I didn't have enough context and assumed that it's common to provide a way to initialize the importer on a "native" library handle. I can remove it if you don't see that being used anywhere.

With the file/data opening I noticed that many plugins prefer to read only from memory, loading external files isn't a problem for ufbx as long as you provide `ufbx_load_opts.filename` or a custom file resolver callback. I thought that loading from files directly would be a bit more memory efficient as ufbx is pretty good at not having to buffer the whole input file during reading (ie. it doesn't need to construct a DOM for the file first).

ufbx can also read from arbitrary user streams so if we want to avoid `<stdio.h>` `FILE`s and there's some `Corrade` file stream I can convert it to use that, or just read always from memory if streaming is not that important.

Reply:
True, I'm guilty of taking a bit of a YOLO approach for file reading, in that if memory use becomes a problem, I'm usually telling people to mmap + `openMemory()` instead. Which means there isn't any Corrade file stream thingy at the moment either.

Since you actually *do* have file reading memory efficient, let's keep using it, then :) But drop the `doOpenState()` and turn it into some `openInternal()` helper that can't be called from outside.

Other usual concerns I have with 3rd party libraries doing file reading on their own is lack of Unicode support on Windows, but you seem to have that implemented as well so all fine.

Reply:
Gotcha, I'll hide the internal open function. I'm on the same boat with not liking libraries that go reading files (which is why ufbx for example doesn't even try to load .mtl files by default, you have to request `ufbx_load_opts.allow_external_files`). Unicode-wise ufbx should be safe though as there's a test case called `synthetic_aβカ😂_7500_ascii.fbx` that is loaded on all platforms.